contrib/aws/aws-sdk-go-v2/aws/aws.go

```diff
@@ -0,0 +1,146 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const (
+	tagAWSAgent     = "aws.agent"
+	tagAWSService   = "aws.service"
+	tagAWSOperation = "aws.operation"
+	tagAWSRegion    = "aws.region"
+	tagAWSRequestID = "aws.request_id"
+)
+
+// AppendMiddleware takes the aws.Config and adds the Datadog tracing middleware into the APIOptions middleware stack.
+// See https://aws.github.io/aws-sdk-go-v2/docs/middleware for more information.
+func AppendMiddleware(awsCfg *aws.Config, opts ...Option) {
+	cfg := &config{}
+
+	defaults(cfg)
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	tm := traceMiddleware{cfg: cfg}
+	awsCfg.APIOptions = append(awsCfg.APIOptions, tm.startTrace, tm.initTrace, tm.deserializeTrace)
+}
+
+type traceMiddleware struct {
+	cfg *config
+}
+
+func (mw *traceMiddleware) startTrace(stack *middleware.Stack) error {
+	return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("TraceStart", func(
+		ctx context.Context, in middleware.InitializeInput, next middleware.InitializeHandler,
+	) (
+		out middleware.InitializeOutput, metadata middleware.Metadata, err error,
+	) {
+		// Start a span with the minimum information possible.
+		// If we get a failure in another init middleware, some context is better than none.
+		opts := []ddtrace.StartSpanOption{
+			tracer.SpanType(ext.SpanTypeHTTP),
+			tracer.ServiceName(serviceName(mw.cfg, "unknown")),
+		}
+		if !math.IsNaN(mw.cfg.analyticsRate) {
+			opts = append(opts, tracer.Tag(ext.EventSampleRate, mw.cfg.analyticsRate))
+		}
+		span, spanctx := tracer.StartSpanFromContext(ctx, "unknown.request", opts...)
```

Comment:
forward a comment from OTel repo, now sure if dd has header based sampler case.
The solution can be record starttime in the first middleware, start the span and update the timestamp in the second middleware. There is an edge case if the error is happened before the span.start you might need to create the span and add error in the first middleware to catch this exceptional case. I will update the PR in OTel repo.

Reply:
I would need a steer from @gbbr on this one.

Reply:
Sorry but I'm not sure I understand the question.

Reply:
To make sure the header based sampler works, we cannot change segment name after it is created.